[ui] Show a run timeline on backfill details pages

bengotow opened the pull request:
## Summary & Motivation

This PR adds a second "Runs" tab to the existing backfill detail page, which presets both list and timeline views of the runs launched by the backfill.

This feature is not behind a feature flag, since it's fairly low risk and available as an additional tab (not the default tab of this view).

<img width="1721" alt="image" src="https://github.com/dagster-io/dagster/assets/1037212/f71aea64-255d-43f4-a887-afa8e0050fe6">


## How I Tested These Changes

salazarm commented on js_modules/dagster-ui/packages/ui-core/src/instance/backfill/BackfillRunsTab.tsx:
```diff
@@ -0,0 +1,187 @@
+    return [startTimestamp * 1000, endTimestamp ? endTimestamp * 1000 : Date.now() + ONE_WEEK];
+  }, [startTimestamp, endTimestamp]);
+
+  const runsForTimelineRet = useRunsForTimeline(timelineRange, filter);
```
Hmm the indexeddb caching doesnt take filtering into account. Think we could add that here? Or maybe we just skip the indexeddb caching entirely for this page?

bengotow replied:
Ooh interesting that's good to know, I think that explains what I was seeing!

bengotow commented:
Hey @salazarm could I get your eyes on this again when you get a chance? 

I reworked the use of your hooks and added filter support to `useRunsForTimeline`. My solution was to give the hourly cache both an ID (the ID of the LRU cache, one per deployment), and a key string (unique to the timeline view you're viewing).

I raised the max number of keys in the LRU cache to 3 so that you can have the main hourlyData key (no name change there!) and then up to two cached backfill timelines:

<img width="1728" alt="image" src="https://github.com/dagster-io/dagster/assets/1037212/26705b4f-883f-46dc-87b0-210704b28a3b">

bengotow commented on js_modules/dagster-ui/packages/ui-core/src/runs/useRunsForTimeline.tsx:
```diff
@@ -443,7 +448,7 @@ export const useRunsForTimeline = ({
   const initialLoading =
     (loadingOngoingRunsData && !ongoingRunsData) ||
-    (loadingCompletedRunsData && !completedRunsQueryData) ||
+    (loadingCompletedRunsData && !completedRunsData) ||
```
I think that this was a bug - since the querydata is present with `{data: undefined, loading: true}`, the completed runs data wasn't incorporated into the loading state really

salazarm replied:
oops good catch

salazarm commented on js_modules/dagster-ui/packages/ui-core/src/runs/useRunsForTimeline.tsx:
```diff
@@ -251,11 +253,13 @@ export const useRunsForTimeline = ({
     const queryData = await client.query<FutureTicksQuery, FutureTicksQueryVariables>({
       query: FUTURE_TICKS_QUERY,
-      variables: {tickCursor: startSec, ticksUntil: _end / 1000.0},
+      variables: showTicks
+        ? {tickCursor: startSec, ticksUntil: _end / 1000.0}
+        : {tickCursor: startSec, ticksUntil: startSec},
```
is the `tickCursor === ticksUntil` to make this effectively no-op? I think we should just skip this call entirely in this case.

bengotow replied:
Yeah agreed - I tried to do that initially, but it turns out that this query also loads workspace / pipeline information that is required down below.  I think no-oping just the ticks part this way works pretty well, but we could also break this out into another query and then skip? I didn't want to impact the main run timeline if I could help it.

salazarm commented on js_modules/dagster-ui/packages/ui-core/src/runs/useRunsForTimeline.tsx:
```diff
@@ -60,9 +62,9 @@ export const useRunsForTimeline = ({
     () =>
       new HourlyDataCache<RunTimelineFragment>(
         localCacheIdPrefix ? `${localCacheIdPrefix}-useRunsForTimeline` : false,
+        JSON.stringify(filter),
```
Hmm, this is weird to me, I think the filtered version should live in a separate cache because I don't want it to evict the main Run timeline from the cache

bengotow replied:
Hmm I like that, I can put ALL of the filtered ones into the same second cache, just didn't want to make a separate cache for each separate filters since nothing ever deletes the indexdbs.

bengotow replied:
Yeah good call, I broke this apart so that if `filter` is present it uses a `useRunsForTimeline-filtered` key instead of interfering with the main one. 

Also verified that the limit=3 is working on the new one and visiting a 4th backfill evicts the oldest item from the LRU cache 🙌 

<img width="1728" alt="image" src="https://github.com/dagster-io/dagster/assets/1037212/ae495efc-189c-4f20-8a91-0b6436f0e147">


<img width="1728" alt="image" src="https://github.com/dagster-io/dagster/assets/1037212/520cfda9-dac1-4100-bd50-c94d73a5d70a">

salazarm replied:
I don't see that in the logic above.

salazarm commented on js_modules/dagster-ui/packages/ui-core/src/runs/HourlyDataCache/HourlyDataCache.tsx:
```diff
@@ -9,16 +9,24 @@ type Subscription<T> = (data: T[]) => void;
+      this.indexedDBCache = cache<string, typeof this.cache>({
         dbName: `HourlyDataCache:${id}`,
-        maxCount: 1, // We only store 1 entry
+        maxCount: 3, // >1 because we have both the overview timeline and backfill timelines
```
I think we should have a separate LRU cache for the backfills page so that it doesn't evict the main Run Timeline cache

salazarm commented on js_modules/dagster-ui/packages/ui-core/src/runs/__stories__/RunTimeline.stories.tsx:
```diff
@@ -35,7 +35,7 @@ export const OneRow = () => {
   }, [sixHoursAgo, now]);
 
-  return <RunTimeline jobs={jobs} range={[sixHoursAgo, now]} />;
+  return <RunTimeline jobs={jobs} rangeMs={[sixHoursAgo, now]} />;
```
Thanks for renaming!

salazarm reviewed (approved):
Mind adding a test case to `useRunsForTimeline` to show that the caches are isolated when you pass a filter?
